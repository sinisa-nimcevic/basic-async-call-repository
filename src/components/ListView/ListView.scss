.list-view {
  max-width: 80%;
  padding: 20px;
  ul {
    margin: 0 0 0 0;
    padding: 0 0 0 0;
  }
}
