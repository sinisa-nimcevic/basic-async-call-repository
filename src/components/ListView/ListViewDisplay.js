import React from 'react';
import ListItem from '../ListItem/ListItem';
import slugify from 'slugify';

const ListViewDisplay = ({ data }) => {
  return (
    <ul className="list-view">
      {data.map((item) => (
        <ListItem item={item} key={`${item.id}-${slugify(item.title)}`} />
      ))}
    </ul>
  );
};

export default ListViewDisplay;
