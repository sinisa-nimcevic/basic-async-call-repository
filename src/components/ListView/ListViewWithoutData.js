import React from 'react';
import slugify from 'slugify';

const ListViewWithoutData = ({ data }) => {
  return (
    <ul className="list-view">
      {data.map((item) => (
        <li key={`${item.id}-${item.title}`}>
          <span>{item.id}</span> {slugify(item.title)}
        </li>
      ))}
    </ul>
  );
};

export default ListViewWithoutData;
