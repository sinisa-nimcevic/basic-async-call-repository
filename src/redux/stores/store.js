import { configureStore } from '@reduxjs/toolkit';
import projectDataReducer from '../slices/projectDataSlice';

export default configureStore({
  reducer: {
    projectData: projectDataReducer,
  },
});
