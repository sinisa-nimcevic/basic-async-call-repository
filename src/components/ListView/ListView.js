import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { getProjectDataSlice } from '../../redux/slices/projectDataSlice';
import { fetchProjectDataThunk } from '../../redux/thunks/projectDataThunk';
import ListViewDisplay from './ListViewDisplay';
import './ListView.scss';

const ListView = () => {
  const dataStatus = useSelector(getProjectDataSlice);
  const { data, loading, error } = dataStatus;
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchProjectDataThunk());
  }, [dispatch]);

  return (
    <div className="list-view">
      <h3>List view</h3>
      {loading && <>loading...</>}
      {error && <>error</>}
      <ListViewDisplay data={data} />
    </div>
  );
};

export default ListView;
