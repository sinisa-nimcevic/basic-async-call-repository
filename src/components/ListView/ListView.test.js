import { fireEvent, render, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import store from '../../redux/stores/store';
import ListView from './ListView';

expect.extend({ toMatchDiffSnapshot });

const dummyItems = [
  {
    id: 1,
    title: 'Top',
    isActive: false,
  },
  {
    id: 2,
    title: 'Kek',
    isActive: false,
  },
  {
    id: 3,
    title: 'Test',
    isActive: false,
  },
];

jest.mock('../../services/dataService', () => {
  return {
    getMyData: () => {
      return {
        data: dummyItems,
        error: null,
      };
    },
  };
});

const renderWrapped = (children) => {
  return render(<Provider store={store}>{children}</Provider>);
};

describe('GIVEN ListView', () => {
  describe('WHEN component is mounted', () => {
    it('THEN renders all of the dummy items', async () => {
      const { getByText } = renderWrapped(<ListView />);

      for (let i = 0; i < dummyItems.length; i++) {
        await waitFor(() => {
          expect(getByText(dummyItems[i].title)).toBeInTheDocument();
        });
      }
    });
  });

  describe('WHEN an item is clicked', () => {
    it('THEN becomes active', async () => {
      const { findByTestId, getByTestId } = renderWrapped(<ListView />);
      const testid = `list-item-display-${dummyItems[1].id}`;

      const testItem = await findByTestId(testid);
      const snapshotA = testItem.cloneNode(true);
      fireEvent.click(testItem);

      expect(snapshotA).toMatchDiffSnapshot(getByTestId(testid));
    });
  });
});
