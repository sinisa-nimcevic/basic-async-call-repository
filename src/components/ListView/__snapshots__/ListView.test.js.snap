// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN ListView WHEN an item is clicked THEN becomes active 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,7 +1,7 @@
  <li
-   class=\\"list-item\\"
+   class=\\"list-item list-item--active\\"
    data-testid=\\"list-item-display-2\\"
  >
    <span
      class=\\"list-item__id\\"
    >"
`;
